### PR TITLE
Retains part of meaningful filename, added unit test and did a small refactoring

### PR DESCRIPTION
**Pull request by Zillode**
Improvements over https://github.com/syncthing/syncthing/pull/1788
Note: all platforms will have a .tmp extension for temporary files (this made unit test and refactoring easier)

### CODE REVIEW

File: internal/model/tempnameutil.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+	return strings.HasPrefix(filepath.Base(name), t.prefix)
+}
+
+func (t tempNamer) TempName(name string) string {
+	tdir := filepath.Dir(name)
+	tbase := filepath.Base(name)
+	if len(tbase) > 100 {
+		hash := md5.New()
+		hash.Write([]byte(name))
+		tbase = fmt.Sprintf("%s%s", tbase[0:100], hash.Sum(nil))
```
Comment (calmh) [inline internal/model/tempnameutil.go]: %s here for hash.Sum ought to be random binary crap which might not be acceptable? Probably want %x?

File: internal/model/tempnameutil.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
+// Copyright (C) 2014 The Syncthing Authors.
```
Comment (calmh) [inline internal/model/tempnameutil.go]: 2015, since this anyway needs a tweak :)

File: internal/model/tempnameutil.go (hunk 1 above)
Comment (calmh) [inline internal/model/tempnameutil.go]: Maybe a separator for clarity? `"%s-%x"`?

**Comment by calmh** [comment]
Maybe skip the util file and drop everything back into just tempname.go with an if on runtime.GOOS for defTempName since that's the only thing that differs now?

File: internal/model/tempnameutil.go (hunk 1 above)
Comment (calmh) [inline internal/model/tempnameutil.go]: Actually, I think this needs either more work or less. If the filename is a minor Russian novel, as in the original issue, the `tbase[0:100]` will more likely than not cut off the last codepoint somewhere in the middle, creating an invalid UTF-8 string. So we either need to treat the string as runes and iterate over it to get a whole number of code points, or just say fuck it and skip the name completely once it's too long. I'd suggest just letting the name through up to 240 characters (as we add 15), and once the length exceeds that just use the hash and nothing else.

**Comment by Zillode** [comment]
@st-jenkins run integration tests for windows

**Comment by calmh** [comment]
:100:

**Pull request merged**